### PR TITLE
[ROCm] enable ROCm on gpu compiler and kernel launch

### PR DESCRIPTION
**Pull request by i-chaochen**
This is a pending and approval PR from https://github.com/tensorflow/tensorflow/pull/59107 that has fixed openXLA on ROCm kernel launch, gpu_executable, and used gpu_version (based on https://github.com/tensorflow/tensorflow/commit/ebbca7d6ee7fc336a8a99f4cccfe1edce309a6d0) to make gpu_compiler more platform-indepdent at the runtime on gemm_rewriter, tree_reduction_rewriter and gemm_rewriter_triton.

@tpopp  @ddunl 

Thanks in advance!

### CODE REVIEW

**Comment by ddunl** [comment]
Thanks for your PR! I see @cheshire and @akuegel were looking at this before, I'll have them take a look.

**Comment by i-chaochen** [comment]
Hi @sherhut it's been over one week this PR without update, wondering will you suggest a reviewer to us, please?

Thanks in advance!

File: xla/service/gpu/gpu_executable.cc [hunk 1]
```diff
@@ -136,6 +137,15 @@ GpuExecutable::GpuExecutable(GpuExecutable::Params params)
...
       constants_(std::move(params.constants)),
       output_info_(std::move(params.output_info)) {
+#if TENSORFLOW_USE_ROCM
+  // ROCm uses hsaco hashes to distinguish between modules.
+  // Bad things happen if multiple modules with identical code are loaded.
+  binary_.resize(binary_.size() + 16);
+  *(uint64_t*)(&binary_[binary_.size() - 16]) = tsl::EnvTime::NowNanos();
+  *(uint64_t*)(&binary_[binary_.size() - 8]) = tsl::random::New64();
+  // workaround for a bug in ROCm 3.3 hipModuleLoadData
+  binary_.reserve(binary_.size() + 256);
```
Comment (akuegel) [inline xla/service/gpu/gpu_executable.cc:147]: What about moving this up before binary_.resize in line 143?
Make it binary_.reserve(binary_.size() + 272);
to reserve both the extra size needed, and the reserve for the workaround.

File: xla/service/gpu/gemm_rewriter_triton_test.cc [hunk 2]
```diff
@@ -57,7 +57,8 @@ ENTRY e {
     lhs_contracting_dims={1}, rhs_contracting_dims={0}
 })")
                     .value();
-  EXPECT_TRUE(GemmRewriterTriton({se::CudaComputeCapability::AMPERE, 0})
+  GpuVersion gpu_version;
```
Comment (akuegel) [inline xla/service/gpu/gemm_rewriter_triton_test.cc:60]: It looks to me that this is a non-functional change. At least I don't see that GpuVersion is automatically initialized to AMPERE, 0.

File: xla/service/gpu/gemm_rewriter_triton_test.cc [hunk 3]
```diff
@@ -82,7 +83,8 @@ ENTRY e {
     lhs_batch_dims={0}, rhs_batch_dims={0}
 })")
                     .value();
-  EXPECT_TRUE(GemmRewriterTriton({se::CudaComputeCapability::AMPERE, 0})
+  GpuVersion gpu_version;
```
Comment (akuegel) [inline xla/service/gpu/gemm_rewriter_triton_test.cc:86]: It looks to me that this is a non-functional change. At least I don't see that GpuVersion is automatically initialized to AMPERE, 0.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/service/gpu/gpu_executable.cc (hunk 1 above)
Comment (i-chaochen) [reply xla/service/gpu/gpu_executable.cc]: Thanks for the suggestion, I have removed that part and simplified as one.

File: xla/service/gpu/gemm_rewriter_triton_test.cc (hunk 3 above)
Comment (i-chaochen) [reply xla/service/gpu/gemm_rewriter_triton_test.cc]: Much appreciate for the review! This change is for gemm_rewriter_triton's GpuVersion, which makes it more genereal and easier to extend in the future.

I have changed the test part to initialize ```{se::CudaComputeCapability::AMPERE, 0}```.

**Review by akuegel — APPROVED** [review]

**Pull request closed without merging**